src/truncate.rs

```diff
@@ -0,0 +1,210 @@
+        builder: Arc<B>,
+        truncate_params: &TruncateQueueParameter,
+    ) -> crate::Result<()> {
+        let origin_name = path.to_str().unwrap();
+        let truncate_file_name = origin_name.to_owned() + "_truncated";
+
+        let buf = PathBuf::from(truncate_file_name);
+        let fd = Arc::new(LogFd::create(buf.as_path())?);
+        let mut log_writer = build_file_writer(builder.as_ref(), buf.as_path(), fd, true)?;
+
+        let mut log_reader =
+            build_file_reader(builder.as_ref(), path, Arc::new(LogFd::open(path)?)).unwrap();
+
+        //write to file
+        self.flush_to_new_file::<B>(&mut log_writer, &mut log_reader, truncate_params)?;
+
+        //rename to original file name
+        fs::rename(buf.as_path(), path)?;
+        Ok(())
+    }
+}
+
+impl TruncateMachine {
+    fn flush_to_new_file<B: FileBuilder>(
+        &mut self,
+        log_writer: &mut LogFileWriter<B>,
+        log_reader: &mut LogFileReader<B>,
+        truncate_params: &TruncateQueueParameter,
+    ) -> crate::Result<()> {
+        let mut batch = LogBatch::default();
+        let mut region_id_to_index = HashMap::<u64, Vec<(Vec<EntryIndex>, Vec<Vec<u8>>)>>::new();
+
+        for log_item_batch in self.items.iter() {
+            for item in log_item_batch.iter() {
+                let item_type = &item.content;
+                let raft_id = item.raft_group_id;
+
+                match item_type {
+                    EntryIndexes(entry_indexes) => {
+                        let mut cursor = 0;
+                        let mut entrys = Vec::new();
+                        while cursor < entry_indexes.0.len() {
+                            let entry = log_reader
+                                .read(entry_indexes.0[cursor].entries.unwrap())
+                                .unwrap();
+                            entrys.push(entry);
+                            cursor += 1;
+                        }
+                        if let Some(v) = region_id_to_index.get_mut(&raft_id) {
+                            v.push((entry_indexes.0.clone(), entrys));
+                        } else {
+                            let vec = vec![(entry_indexes.0.clone(), entrys)];
+                            region_id_to_index.insert(raft_id, vec);
+                        }
+                    }
+                    Command(cmd) => batch.add_command(raft_id, cmd.clone()),
+                    Kv(kv) => match kv.op_type {
+                        log_batch::OpType::Put => {
+                            batch.put(raft_id, kv.key.clone(), kv.value.as_ref().unwrap().to_vec())
+                        }
+                        log_batch::OpType::Del => batch.delete(raft_id, kv.key.clone()),
+                    },
+                }
+            }
+        }
+
+        let raft_group_ids = truncate_params.raft_groups_ids;
+        let truncate_mode = &truncate_params.truncate_mode;
+
+        for (k, v) in region_id_to_index.iter() {
```

Comment:
I prefer to do the "corruption detection" inside `replay` function. Most of the file might not need rewriting at all. Also, you can assume there is no corruption inside one log item. This might help you simplify the code.

Reply:
"corruption detection" logic can indeed do inside `replay` function. The reason that i do not do it is we lack the essential parameters in the method for example `raft_groups_ids` and `truncate_mode`, Changing method `replay` signature is costly.   About this problem,  do you have some ideas?

Reply:
As I commented earlier, you should pass in those necessary information when calling `TruncateMachine::new`.

Reply:
Currently TruncateMachine and other struct that implements `ReplayMachine` was created by `default` method, how do we pass other information without change the `default` method? 
https://github.com/tikv/raft-engine/blob/89c3f39911b38dd2dcc9f9aa348187cc73ca1aaa/src/file_pipe_log/pipe_builder.rs#L221-L225

Reply:
You are right, I forgot about this. Maybe pass in a replay machine factory:
```
pub trait ReplayMachineFactory<M: ReplayMachine> {
  fn new_machine(&self) -> M;
}
```

Reply:
This can create `ReplayMachine`  instance indeed, but i'm afraid that we should still add params `TruncateQueueParameter` which comes from caller to `new_machine` method to create `TruncateMachine` instance,  It's seems to be not very elegant. 

In other words, if we do not add a `truncate` method to replace `recover` ,  information about truncate action have to be given in `recovery` which is not very desirable. So what about add a new `truncate` method?

Reply:
You can pass the information to factory.

Reply:
@tabokie 
![image](https://user-images.githubusercontent.com/15794564/146484167-31a881db-b30d-4b74-af9e-c7a87b832a2f.png)

Is this elegant enough,  if it's OK, i will continue my work

Reply:
No it's still the same, What I meant before is you can pass in an actual factory (not just a generic type), which contains all the information.